README.md
halocksmith
===========
[![Build Status](https://secure.travis-ci.org/danmactough/halocksmith.png?branch=master)](https://travis-ci.org/danmactough/halocksmith)

Distributed locking mechanism using [haredis](https://github.com/carlos8f/haredis)

This module is a fork of Ian Hansen's [locksmith](https://github.com/supershabam/locksmith).
It implements the locking algorithm described in the Redis documentation: http://redis.io/commands/setnx

## Purpose

Enables locking (mutex/semaphore) across a clustered node application.

## API
```javascript
/**
 * Options:
 *
 * nodes {Array} (optional): host/port array; defaults to ['127.0.0.1:6379']; See [haredis](https://github.com/carlos8f/haredis#createclient) for more information
 * prefix {String} (optional): defaults to '__halocksmith:'
 * timeout {Number} (optional): given in seconds; defaults to 120 (two minutes)
 * retries {Number} (optional): number of times to retry acquiring the lock; defaults to 100
 * retryTimeout {Number} (optional): given in milliseconds; defaults to 1000 (one second)
 * redisClient {redisClient} (optional): already instantiated Redis client (other Redis options won't be used)
 * **any additional options are passed to redis.createClient as options**
 */
var lock = require('halocksmith')(options);

/**
 * lock is a function
 *
 * key {String} (optional): the cluster-wide keyname to lock on; defaults to ''
 * callback {Function}: function to execute when you have the lock; takes two parameters:
 *     - error {Error}
 *     - release {Function}: function to call when you are done with the lock (required in most circumstances); takes an optional callback
 */
lock([key], callback)
```

Example
-------

```javascript
var lock = require('halocksmith')({
  timeout: 120
});

// `key` may be omitted if you only need one lock
lock(function(err, release) {
  // Return without releasing the lock if there's an error
  if (err) return console.error(err);

  // This is the only process in the cluster that acquired the lock
  doSomething();

  // Delete the lock; other processes will be able to acquire a new lock
  release(function (err) {
    // optional callback
    if (err) return console.error(err);
    doSomeOtherThing();
  });
});
```
- - -

### Developed by [Terra Eclipse](http://www.terraeclipse.com)
Terra Eclipse, Inc. is a nationally recognized political technology and
strategy firm located in Aptos, CA and Washington, D.C.

- - -

### License: MIT
Copyright (c) 2013 Dan MacTough ([http://yabfog.com](http://yabfog.com))

Copyright (c) 2013 Terra Eclipse, Inc. ([http://www.terraeclipse.com](http://www.terraeclipse.com))

Portions Copyright (c) 2012, Ian Hansen (//github.com/supershabam)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is furnished
to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


History.md

v0.0.3 / 2013-04-05
==================

  * Add travis-ci
  * Use mocha for tests and add basic test
  * Fix default timeout
  * Change option parsing to handle passing only redis config

v0.0.2 / 2013-04-05
==================

  * Force no slave on read
  * Change release method to take a callback and not log to console
  * Add option to set the retry timeout
  * Update README, package.json, LICENSE
  * Use haredis instead of node-redis
  * Bail as soon as possible if we aren't supposed to retry
  * Fix bug in option parsing where valid options could be zero
  * Fix off-by-one bug in retry that caused one too many retries
  * Don't pass retries option to redis
  * Clarify code comment


test/basic.js
describe('halocksmith', function () {
  var cleanup;

  after(function (done) {
    cleanup(done);
  });

  it('can create an exclusive lock', function (done) {
    var lock = halocksmith({ retries: 0, timeout: 2 });
    lock(function (err, release) {
      assert.ifError(err);
      cleanup = release;
      setTimeout(done, 100);
    });
    lock(function (err, release) {
      assert.ok(err instanceof Error);
      assert.equal(release, null);
    });
  });
});

test/common.js
/*global assert: true, halocksmith: true*/

halocksmith = require('../');
assert = require('assert');


index.js
'use strict';

/**
 * The locking algorithm used in this is described here: http://redis.io/commands/setnx
 */
var redis = require('haredis')
  , moment = require('moment')
  ;

function HALocksmith(options) {
  var nodes, split, prefix, timeout, retries, retryTimeout;

  // Allow passing just the redis config as options
  if (typeof options === 'string') {
    split = options.split(':');
    options = {host: split[0]};
    if (split[1] && split[1].match(/^\d+$/)) {
      options.port = parseInt(split[1], 10);
    }
  }
  else if (typeof options === 'number') {
    options = {port: options};
  }
  else if (options && !Array.isArray(options) && options['0']) {
    // Support for nodes as object, like optimist might produce.
    nodes = [];
    Object.keys(options).forEach(function (k) {
      if (k.match(/^\d+$/)) {
        nodes.push(options[k]);
        delete options[k];
      }
    });
    options.nodes = nodes;
  }
  else if (Array.isArray(options)) {
    options = {nodes: options};
  }
  // Full option parsing
  else {

    options = options || { host: '127.0.0.1', port: 6379 };
    if (options.port && !options.host) options.host = '127.0.0.1';
    prefix = options.prefix || '__halocksmith:';
    timeout = 'timeout' in options ? options.timeout : 120;
    retries = 'retries' in options ? options.retries : 100;
    retryTimeout = 'retryTimeout' in options ? options.retryTimeout : 1000;

    delete options.prefix;
    delete options.timeout;
    delete options.retries;
    delete options.retryTimeout;

  }

  if (options.redisClient) {
    this._redisClient = options.redisClient;
  } else if (options.nodes) {
    // haredis node list
    this._redisClient = redis.createClient(options.nodes, options);
  }
  else {
    this._redisClient = redis.createClient(options.port, options.host, options);
  }

  this._prefix = prefix;
  this._timeout = timeout;
  this._retries = retries;
  this._retryTimeout = retryTimeout;
}

HALocksmith.prototype.lock = function(key, cb) {
  var retries = 0
    , _this = this
    , fullKey
    ;

  if (typeof(key) === 'function') {
    cb = key;
    key = '';
  }
  fullKey = this._prefix + key;

  (function aquire() {
    var expires = moment().add('seconds', _this._timeout).unix();
    _this._redisClient.setnx(fullKey, expires, function handleSetnx(err, response) {
      if (err) return cb(err);

      // if we aquired the lock
      if (response === 1) {
        return cb(null, _this._release.bind(_this, key, expires));
      }

      // otherwise let's check if the lockholder is expired
      _this._redisClient.slaveOk(false).get(fullKey, function handleGet(err, keyExpires) {
        if (err) return cb(err);

        // no retrying allowed and the key doesn't exist, so there is no deadlock to try to fix
        if (!_this._retries && !keyExpires) {
          return cb(new Error('failed to acquire lock for: ' + key));
        }

        function retry() {
          if (++retries > _this._retries) {
            return cb(new Error('maximum retries hit while aquiring lock for: ' + key));
          }
          setTimeout(aquire, _this._retryTimeout);
        }

        // if the key still exists and has not expired
        if (moment().unix() < keyExpires) {
          return retry();
        } else { // try and aquire expired lock
          expires = moment().add('seconds', _this._timeout).unix();
          _this._redisClient.getset(fullKey, expires, function handleGetSet(err, keyExpires) {
            if (err) return cb(err);

            // if the key is no longer expired, somebody else grabbed it, get back in line
            if (moment().unix() < keyExpires) {
              return retry();
            }

            // we got the lock!
            return cb(null, _this._release.bind(_this, key, expires));
          });
        }
      });
    });
  })();
};

HALocksmith.prototype._release = function(key, expires, callback) {
  var fullKey = this._prefix + key;

  // callback is optional
  if ('function' !== typeof callback) {
    callback = function () {};
  }

  // nice! we finished before somebody tries to expires us
  if (moment().unix() < expires) {
    this._redisClient.del(fullKey, function handleDel(err) {
      if (err) return callback(err);
      callback();
    });
  } else {
    // it's too late, the lock is already being fought for
    callback(new Error('you released your lock after expiration on key: ' + key));
  }
};

/**
 * options = {
 *   nodes {Array} (optional): host/port array; defaults to ['127.0.0.1:6379']; See [haredis](https://github.com/carlos8f/haredis#createclient) for more information
 *   prefix {String} (optional): defaults to '__halocksmith:'
 *   timeout {Number} (optional): given in seconds; defaults to 120 (two minutes)
 *   retries {Number} (optional): number of times to retry acquiring the lock; defaults to 100
 *   retryTimeout {Number} (optional): given in milliseconds; defaults to 1000 (one second)
 *   redisClient {redisClient} (optional): already instantiated Redis client (other Redis options won't be used)
 *   **any additional options are passed to redis.createClient as options**
 * }
 */
module.exports = function(options) {
  var halock = new HALocksmith(options);
  return halock.lock.bind(halock);
};
module.exports.HALocksmith = HALocksmith;
